internal/zinc-apiinfo/src/main/scala/sbt/internal/inc/ClassToAPI.scala

```diff
@@ -548,56 +548,18 @@ object ClassToAPI {
   }.toMap
   def primitive(name: String): api.Type = PrimitiveRefs(name)
 
-  // Workarounds for https://github.com/sbt/sbt/issues/1035
-  //   these catch the GenericSignatureFormatError and return the erased type
-
-  private[this] def returnType(f: Field): Type =
-    try f.getGenericType
-    catch {
-      case _: GenericSignatureFormatError => f.getType
-    }
-  private[this] def parameterTypes(c: Constructor[_]): Array[Type] =
-    try c.getGenericParameterTypes
-    catch {
-      case _: GenericSignatureFormatError => convert(c.getParameterTypes)
-    }
-  private[this] def exceptionTypes(c: Constructor[_]): Array[Type] =
-    try c.getGenericExceptionTypes
-    catch {
-      case _: GenericSignatureFormatError => convert(c.getExceptionTypes)
-    }
-  private[this] def parameterTypes(m: Method): Array[Type] =
-    try m.getGenericParameterTypes
-    catch {
-      case _: GenericSignatureFormatError => convert(m.getParameterTypes)
-    }
-  private[this] def returnType(m: Method): Type =
-    try m.getGenericReturnType
-    catch {
-      case _: GenericSignatureFormatError => m.getReturnType
-    }
-  private[this] def exceptionTypes(m: Method): Array[Type] =
-    try m.getGenericExceptionTypes
-    catch {
-      case _: GenericSignatureFormatError => convert(m.getExceptionTypes)
-    }
-
-  private[this] def typeParameterTypes[T](m: Constructor[T]): Array[TypeVariable[Constructor[T]]] =
-    try m.getTypeParameters
-    catch {
-      case _: GenericSignatureFormatError => new Array(0)
-    }
-  private[this] def typeParameterTypes[T](m: Class[T]): Array[TypeVariable[Class[T]]] =
-    try m.getTypeParameters
-    catch {
-      case _: GenericSignatureFormatError => new Array(0)
-    }
-  private[this] def typeParameterTypes(m: Method): Array[TypeVariable[Method]] =
-    try m.getTypeParameters
-    catch {
-      case _: GenericSignatureFormatError => new Array(0)
-    }
-
-  private[this] def convert(classes: Array[Class[_]]): Array[Type] =
-    classes.asInstanceOf[Array[Type]] // ok: treat Arrays as read-only
+  @inline private[this] def returnType(f: Field): Type = f.getGenericType
```

Comment:
Inlining `getGenericType` is probably safe, https://github.com/scala/bug/issues/6789 a special case for `forName`. Note that, since 2.12.3, you have to be explicit in telling the inliner where it's allowed to inline from (https://github.com/scala/scala/pull/5964).

Inlining from the JDK is probably not a good idea in general as the bytecode potentially becomes non-portable. There might be licensing issues too. Maybe that's not a problem here because that code is only compiled locally for every user and never distributed in binary.

`@inline` is not a hint to the heuristic, but it makes the optimizer inline unconditionally (if the method can be inlined). The heuristic applies for every method invocation.

Marking small forwarder methods `@inline` is usually not beneficial, the JVM handles small monomorphic methods just fine. The only potential issue with many small methods on a hot path is that they consume from the `-XX:MaxInlineLevel` budget.

Inlining in Scala has benefits when it can make the code easier for the VM to optimize, in particular when callsites become monomorphic or escape analysis / stack allocation starts working.

> it's a good practice to use @inline where applicable

I don't agree here. For small methods like here there's no gain. For bigger methods, there's a risk of slowdown due to code size. Good practice is to use the annotation where performance benefits are measured (or known).

I once made a small experiment and removed all `@inline` annotations from the Scala compiler, and performance didn't change. It might be different now, we'd have to do it again. But I think the heuristics cover the main cases. As another data point, there's a stress test mode in the optimizer `-Yopt-inline-heuristics:everything`. If you build a compiler with that, it runs 2x slower.

In general when working on JVM performance optimization, measuring is really important. More often than not, just reasoning about it leads to changes that don't improve or even hurt performance - there are just too many factors in play.